### ISSUE TITLE
Upgrade the mod dependency of Prometheus

### ISSUE THREAD

**Issue by liguozhong** [prometheus-operator/prometheus-operator#3380]
https://github.com/prometheus/prometheus/releases/tag/v2.19.3

**Comment by christiangda** [comment]
Is the version [v2.20.0](https://github.com/prometheus/prometheus/releases/tag/v2.20.0) already supported?

**Comment by brancz** [comment]
Feel free to send us a PR :)

**Comment by lilic** [comment]
@christiangda  Version 2.20 of promethus is supported by the Prometheus operator.
https://github.com/coreos/prometheus-operator/blob/566b18b2c9bf62ff3558804a69de5e1127ce8171/Documentation/compatibility.md#prometheus

or are you also asking about bumping the dependency in go.mod like the original requester?

In any case PRs for bumping go.mod if we have not yet, are very welcome! 🎉

**Comment by christiangda** [comment]
Hi @lilic and @brancz 

Since yesterday I'm trying  to update/upgrade `prometheus-operator` dependencies and guided by versions on [prometheus go.mod](https://github.com/prometheus/prometheus/blob/master/go.mod) without any success.

When I tried to use the latest version of prometheus project dependency many of the project dependencies broke.  Of course, nothing new for you, I supposed you are up to date with this.

I update the `pormehteus package dependency` using the command:

```bash
go get -d github.com/prometheus/prometheus@v2
``` 

Output:
```bash
go: github.com/prometheus/prometheus v2 => v2.5.0+incompatible
```

Many of the project files have problems with this dependency version because they refactor a lot of packages in the version " v2.5.0+incompatible".

The curious thing for me was when I tried to update the `prometheus package dependency` using the command:

```bash
go get -d github.com/prometheus/prometheus@v2.20.0
``` 
and get the output:

```bash
go get github.com/prometheus/prometheus@v2.20.0: github.com/prometheus/prometheus@v2.20.0: invalid version: module contains a go.mod file, so major version must be compatible: should be v0 or v1, not v2
```

and trying to understand why I can't get the `prometheus project dependency` as versions they public, I found this haunting issue reported: https://github.com/prometheus/prometheus/issues/6048 

And what is clear, `prometheus project` doesn't follow any `go packages good practice` and this is the reason why is too hard to use as a dependency.

Of course, all the mortals like me use the `standard got library` published by them on [client_golang](https://github.com/prometheus/client_golang) to instrument my apps, but `prometheus-operator project` needs the `"internals packages"` of `prometheus project` so:

Are any kind of conversations between the `prometheus-operator team` and `prometheus project team` in order to clarify how `prometheus project team` is going to export `pkg or/and internal folders` packages in order to allow other projects like this (prometheus-operator) used these?

**Comment by brancz** [comment]
We tend to pin to specific commits or just `@master`. That so far has worked out just fine. Some of us are Prometheus maintainers, so yes that conversation does happen, but there is no conclusion yet, there are a couple of paths that the Prometheus project may go down and it's not decided yet.

**Comment by christiangda** [comment]
Hi @brancz , 

Thank you for your comments about it.  I hope both teams in a short time find a solution that allows mortals like me to help in this project.

I have been using both projects intensively like the majority, and I customized my job here 
[https://github.com/slashdevops/kube-visibility](https://github.com/slashdevops/kube-visibility)

Which is just like [kube-prometheus](https://github.com/prometheus-operator/kube-prometheus) but instead of using `jsonnet` I'm using `kustomize` and [kpt](https://googlecontainertools.github.io/kpt/)

**Comment by brancz** [comment]
You know that we generate a kustomize compatible config as the default in kube-prometheus? :)

**Comment by christiangda** [comment]
@brancz yes I do, in fact, those manifest was what I was using for a long time.  

The problem where when I started to use fluxcd (GitOps Operator) and the job of keeping those manifest updated in an easy way.  Believe me, IO tried with bash scripts (curl the main) into my CI pipeline, etc.   At the end of the day, I found `kept` which is an amazing tool that allows you to establish a contract (like interfaces in golang) between your customization and the package you are customizing.

My best wishes will be that every `kube-prometheus` component (which are the same of kube-visibility) exposes its k8s manifests and customization using [kpt blueprints](https://googlecontainertools.github.io/kpt/guides/producer/blueprint/), like how they are doing right now with `jsonnet` or `helm charts`
